### ISSUE TITLE
NVMeFix redesign

### ISSUE THREAD

**Issue by 07151129** [acidanthera/bugtracker#741]
Once we implement https://github.com/acidanthera/bugtracker/issues/736 to get early patching, we should make the following changes:

- [ ] Synchronous controller detection via controller PCI init hooks. This will simplify our design by reducing the need to synchronise. We should still keep a detected controller list to track state specific to our code.

- [ ] Patches for early IONVMeController init to aid debugging for unsupported controllers. In particular, queue setup (depth, tags),  timeouts, verifying IRQ config, forcing polled I/O.

- [ ] Experimental PCI ASPM feature. It's known to be unstable for some controllers, so turn it off by default.

- [ ] Switch to individual properties instead of quirk mask from Linux, as we don't use most of those bits anyway. Make sure to introduce a name prefix to avoid potential clash with Apple code.

- [x] Disable active NVMe PM by default if APST is enabled. IOPM has 1 s resolution compared to microseconds for APST, so we should only use it as fallback.

- [ ] `IONVMeController::ProcessIdentifyData` does not account for MPSMIN.

- [x] ~~`RequestAsyncEvents` can cause a panic sometimes, can likely be patched out as a last resort.~~ It seems to be no-op as the async event counter is never changed from 0. Regardlessly, we still see it in some panics, and patching it out is reported to help on PM981. With some prolonged testing we expect to verify if it indeed helps.

- [ ] Polled Identify submission may be forced to debug hangs for some controllers.

- [ ] Add CQ polled read (https://lists.ubuntu.com/archives/kernel-team/2018-December/097209.html) as a last resort for IRQ timeout.

**Comment by tiger511** [comment]
> * [x]  ~`RequestAsyncEvents` can cause a panic sometimes, can likely be patched out as a last resort.~ It seems to be no-op as the async event counter is never changed from 0. Regardlessly, we still see it in some panics, and patching it out is reported to help on PM981. With some prolonged testing we expect to verify if it indeed helps.

Does that mean PM981 will be supported? Currently it bootloops while installing.If you have plans to make it suppport,i can provide you with logs and everything.

**Comment by 07151129** [comment]
@tiger511 well, yes, that's what the README says. If it doesn't work for you for some reason, feel free to open a separate issue.

**Comment by sinag** [comment]
I have a PM981, would like to help test this. What should I do?

**Comment by abcslayer** [comment]
Hi, I got a kp with NVMeFix 1.0.3 with PM981 HP version (P/N MZVLB512HAJQ-000H1, FW: EXA72H1Q). The panic came with IONVMeController.cpp: 5320, in the backtrace there is RequestAsyncEvents.
Please check the attached file, hope it is helpful. Thank you.
![IMG_20200612_025557](https://user-images.githubusercontent.com/5021088/84524720-9b961480-ad04-11ea-8dac-de7567cf6fe0.jpg)

**Comment by abcslayer** [comment]
I can confirm the that same NVMe drive always has the same kernel panic in recovery boot, installation boot or macOS boot from external drive. The same drive if was used via RAW disk access in VirtualBox (as a NVMe drive) will be OK.

**Comment by antoniomcr96** [comment]
Same here. 
Impossible to install from scratch. As suggested in other issues I tried to install it in an external drive with a VM and then clone the drive.
It booted the first times, to my surprise, with some random kernel panic (also on shutdown, I noticed that it happened when EFI partition was mounted on the system). Partition map EFI - NTFS (Windows) - EXT4 (Linux) - APFS 
Now for some strange reason it refuses to boot at all. Booting from external drive is ok.
The funny thing is that also booting from external drive sometimes results in IONVMEController kernel panic. Removing the drive from the laptop avoids every panic 

Anyway thanks in advance for all your work!
